gstreamer-source.c

```diff
@@ -308,6 +308,108 @@ const char *gstreamer_source_get_name(void *type_data)
 	return "GStreamer Source";
 }
 
+enum obs_media_state gstreamer_source_get_state(void *user_data)
+{
+	data_t *data = user_data;
+	GstStateChangeReturn ret;
+	GstState state;
+
+	ret = gst_element_get_state(data->pipe, &state, NULL, 0);
+	if (ret == GST_STATE_CHANGE_FAILURE)
+		return OBS_MEDIA_STATE_ERROR;
+
+	switch (state) {
+	case GST_STATE_VOID_PENDING:
+		return OBS_MEDIA_STATE_NONE;
+	case GST_STATE_NULL:
+		return OBS_MEDIA_STATE_OPENING;
+	case GST_STATE_READY:
+		return OBS_MEDIA_STATE_STOPPED;
+	case GST_STATE_PAUSED:
+		return OBS_MEDIA_STATE_PAUSED;
+	case GST_STATE_PLAYING:
+		return OBS_MEDIA_STATE_PLAYING;
+	}
+
+	return 0;
+}
+
+int64_t gstreamer_source_get_time(void *user_data)
+{
+	data_t *data = user_data;
+	int64_t position;
+
+	if (gst_element_query_position(data->pipe, GST_FORMAT_TIME, &position))
+		if (GST_CLOCK_TIME_IS_VALID(position))
+			return GST_TIME_AS_MSECONDS(position);
+
+	return 0;
+}
+
+int64_t gstreamer_source_get_duration(void *user_data)
+{
+	data_t *data = user_data;
+	int64_t duration;
+
+	if (gst_element_query_duration(data->pipe, GST_FORMAT_TIME, &duration))
+		if (GST_CLOCK_TIME_IS_VALID(duration))
+			return GST_TIME_AS_MSECONDS(duration);
+
+	return 0;
+}
+
+void gstreamer_source_play_pause(void *user_data, bool pause)
+{
+	data_t *data = user_data;
+
+	GstState state = pause ? GST_STATE_PAUSED : GST_STATE_PLAYING;
+	gst_element_set_state(data->pipe, state);
```

Comment:
Could have just used the ternary without an extra variable, but I don't mind that either.

Reply:
Dropped the 'state' variable in my [v2].